contracts/filters/MerkleCollateralFilter.sol

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.20;
+
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+import "../CollateralFilter.sol";
+
+/**
+ * @title Merkle Collateral Filter
+ * @author MetaStreet Labs
+ */
+contract MerkleCollateralFilter is CollateralFilter {
+    /**************************************************************************/
+    /* Errors */
+    /**************************************************************************/
+
+    /**
+     * @notice Invalid root
+     */
+    error InvalidRoot();
+
+    /**
+     * @notice Invalid node count
+     */
+    error InvalidNodeCount();
+
+    /**
+     * @notice Invalid context
+     */
+    error InvalidContext();
+
+    /**************************************************************************/
+    /* State */
+    /**************************************************************************/
+
+    /**
+     * @notice Supported token
+     */
+    address private _token;
+
+    /**
+     * @notice Number of nodes for each proof
+     */
+    uint32 private _nodeCount;
+
+    /**
+     * @notice Merkle root
+     */
+    bytes32 private _root;
+
+    /**
+     * @notice Metadata URI
+     */
+    string private _metadataURI;
+
+    /**************************************************************************/
+    /* Initializer */
+    /**************************************************************************/
+
+    /**
+     * @notice MerkleCollateralFilter initializer
+     */
+    function _initialize(address token, bytes memory params) internal {
+        /* Decode parameters */
+        (bytes32 root, uint32 nodeCount, string memory metadataURI_) = abi.decode(params, (bytes32, uint32, string));
+
+        if (root == bytes32(0)) revert InvalidRoot();
+        if (nodeCount == 0) revert InvalidNodeCount();
+
+        _token = token;
+        _root = root;
+        _nodeCount = nodeCount;
+        _metadataURI = metadataURI_;
+    }
+
+    /**************************************************************************/
+    /* Helpers */
+    /**************************************************************************/
+
+    /**
+     * @notice Helper function that returns merkle proof in bytes32[] shape
+     * @param context Context
+     * @param nodeCount Node count
+     * @return merkleProof Merkle proof
+     */
+    function _merkleProof(
+        bytes calldata context,
+        uint256 nodeCount
```

Comment:
We don't really need nodeCount -- context.length must be a multiple for this function to work, so we can just do `context.length / 32` if we're already validating the context length in `_collateralSupported()`.